pennylane/_qubit_device.py

```diff
@@ -156,6 +188,54 @@ def reset(self):
         """
         self._samples = None
 
+    @contextlib.contextmanager
+    def custom_decomp_context(self):
+        """A context manager for applying custom decompositions."""
+
+        from pennylane.transforms.qfunc_transforms import NonQueuingTape
+        NonQueuingTape = type("NonQueuingTape", (NonQueuingTape, qml.tape.QuantumTape), {})
+
+        # Creates an individaul context
+        @contextlib.contextmanager
+        def _custom_decomposition(obj, fn):
+            # Covers the case where the user passes a string to indicate the Operator
+            if isinstance(obj, str):
+                obj = getattr(qml, obj)
+
+            original_decomp_method = obj.decompose
+
+            # This is the method that will override the current .decompose method of
```

Comment:
The reason why we have to override `decompose` instead of `decomposition` here is because in the default operator `expand` function, this is what is called to perform the decomposition (see [here](https://github.com/PennyLaneAI/pennylane/blob/ec0b0487ffdce9c45604f0043f24498cdc6c9de0/pennylane/operation.py#L747)).

Reply:
nice! this is more general as well

Reply:
The one disadvantage I can see is that, if a user wanted to define a custom decomposition for a _template_, this wouldn't work, because many of them don't have a `decomposition`, just `expand`. But this is maybe not so much the intended use case.

Reply:
this will be changed shortly though, right? (@mariaschuld?)